### PR TITLE
Added benefits for one controller per endpoint

### PR DESCRIPTION
**Pull request by richev-o**
The Single API Framework team have adopted a pattern of having one controller for each API endpoint. The PR aims to document the reasons and benefits behind this approach.

### CODE REVIEW

File: OctopusServer/Api/AspNet.md [hunk 1]
```diff
@@ -26,6 +26,14 @@ If an ASP.NET Controller endpoint exists for the HTTP Route + Verb pair coming i
 
 We have opted for a **One Controller Per Endpoint** structure, meaning for each HTTP Route + Verb, a Controller is created to serve that request.
 
+We have found many benefits of this approach (when compared to resource-centric controllers containing methods for many endpoints). Specifically:
+
+* Smaller code
+* Better dependency graph for endpoints, as 
+* Potentially testable endpoints (if we want them) thanks to fewer dependencies per-controller.
+
+Because we are making controllers more explicit and flattening that inheritance heirarchy, it allows them to contain more code without losing clarity. Things would get messy quickly if we had many endpoints per controller, each incliding some additional amount of code.
```
Comment (paulegradie) [inline OctopusServer/Api/AspNet.md:35]: including*

File: OctopusServer/Api/AspNet.md [hunk 2]
```diff
@@ -26,6 +26,14 @@ If an ASP.NET Controller endpoint exists for the HTTP Route + Verb pair coming i
 
 We have opted for a **One Controller Per Endpoint** structure, meaning for each HTTP Route + Verb, a Controller is created to serve that request.
 
+We have found many benefits of this approach (when compared to resource-centric controllers containing methods for many endpoints). Specifically:
+
+* Smaller code
+* Better dependency graph for endpoints, as
```
Comment (paulegradie) [inline OctopusServer/Api/AspNet.md:32]: (as ?)

File: OctopusServer/Api/AspNet.md [hunk 3]
```diff
@@ -26,6 +26,14 @@ If an ASP.NET Controller endpoint exists for the HTTP Route + Verb pair coming i
...
 We have opted for a **One Controller Per Endpoint** structure, meaning for each HTTP Route + Verb, a Controller is created to serve that request.
 
+We have found many benefits of this approach (when compared to resource-centric controllers containing methods for many endpoints). Specifically:
+
+* Smaller code
+* Better dependency graph for endpoints, as 
+* Potentially testable endpoints (if we want them) thanks to fewer dependencies per-controller.
+
+Because we are making controllers more explicit and flattening that inheritance heirarchy, it allows them to contain more code without losing clarity. Things would get messy quickly if we had many endpoints per controller, each incliding some additional amount of code.
+
```
Comment (paulegradie) [inline OctopusServer/Api/AspNet.md:36]: A few more ideas to consider that may be helpful for future folks wondering why we went this road:

 - Could consider adding something more around how we get a more 1:1 relationship with the base class - each endpoint clearly depends on any base class it inherits from.

- Could consider stating that the pattern generally falls in line with the S in SOLID, single responsibility principle - where each controller is responsible for a single thing - servicing a single endpoint. 

 - Could consider a statement around how keeping 1 endpoint/controller, we deliberately prevent ourselves from grouping endpoints. 

 - Could consider how it affects the naming conventions we use -- we are able to name controllers with the method (& pseudomethod) terminology that the controller implements.

File: OctopusServer/Api/AspNet.md (hunk 2 above)
Comment (richev-o) [reply OctopusServer/Api/AspNet.md]: My bad, didn't finish the sentence.

File: OctopusServer/Api/AspNet.md [hunk 4]
```diff
@@ -26,6 +26,14 @@ If an ASP.NET Controller endpoint exists for the HTTP Route + Verb pair coming i
 
 We have opted for a **One Controller Per Endpoint** structure, meaning for each HTTP Route + Verb, a Controller is created to serve that request.
 
+We have found many benefits of this approach (when compared to resource-centric controllers containing methods for many endpoints). Specifically:
+
+* Smaller code
+* Better dependency graph for endpoints, as there are fewer per controller
+* Potentially testable endpoints (if we want them) thanks to fewer dependencies per-controller.
```
Comment (cailyoung) [inline OctopusServer/Api/AspNet.md:33]: Can you explain what you mean by 'potentially testable' here? Unit tests for the controller? I thought we were already testing them! 😄

File: OctopusServer/Api/AspNet.md (hunk 3 above)
Comment (richev-o) [reply OctopusServer/Api/AspNet.md:38]: Thanks - I've addressed _some_ of these.

File: OctopusServer/Api/AspNet.md (hunk 4 above)
Comment (andrewabest) [reply OctopusServer/Api/AspNet.md:33]: Yes, unit testable 👍

File: OctopusServer/Api/AspNet.md [hunk 5]
```diff
@@ -24,7 +24,17 @@ If an ASP.NET Controller endpoint exists for the HTTP Route + Verb pair coming i
 
 ## Architecture
 
-We have opted for a **One Controller Per Endpoint** structure, meaning for each HTTP Route + Verb, a Controller is created to serve that request.
+We have opted for a **One Controller Per Endpoint** structure, meaning for each HTTP Route + Verb, a Controller is created to serve that request. The name of the controller should contain the name of the resource concerned, with a verb that explains what is being done to it (e.g. `Get`, `Create`, `Update`, `Delete`). The public action method on the controller should be named using this same verb.
+
+We have found many benefits of this approach (when compared to resource-centric controllers containing methods for many endpoints). Specifically:
+
+* Smaller code
```
Comment (andrewabest) [inline OctopusServer/Api/AspNet.md:31]:
```suggestion
* The controller only does one thing
```

File: OctopusServer/Api/AspNet.md [hunk 6]
```diff
@@ -24,7 +24,17 @@ If an ASP.NET Controller endpoint exists for the HTTP Route + Verb pair coming i
...
+* Better dependency graph for endpoints, as there are fewer per controller
```
Comment (andrewabest) [inline OctopusServer/Api/AspNet.md:32]:
```suggestion
* Controllers only take the dependencies required to do the one thing they need to do
```

File: OctopusServer/Api/AspNet.md [hunk 7]
```diff
@@ -24,7 +24,17 @@ If an ASP.NET Controller endpoint exists for the HTTP Route + Verb pair coming i
...
+* Potentially testable endpoints (if we want them) thanks to fewer dependencies per-controller.
```
Comment (andrewabest) [inline OctopusServer/Api/AspNet.md:33]:
```suggestion
* Controllers can be unit-tested more easily as they have simpler dependencies
```

File: OctopusServer/Api/AspNet.md [hunk 8]
```diff
@@ -24,7 +24,17 @@ If an ASP.NET Controller endpoint exists for the HTTP Route + Verb pair coming i
...
+
```
Comment (andrewabest) [inline OctopusServer/Api/AspNet.md:34]:
```suggestion
* We make a deliberate break away from controllers _having_ to be resource-aligned, giving more flexibility for controllers to exist that might model aggregate domain operations
```

**Review by andrewabest — APPROVED** [review]
This looks like a good summary of the intent of the architecture. 🚢

**Pull request merged**